### PR TITLE
Improve strategy for selecting targets to be scraped for examples

### LINKED ISSUE #10571: Cargo panics when using `-Z rustdoc-scrape-examples=examples` on proc-macro crates
### Problem

Cargo (nightly) panics when I use `-Z rustdoc-scrape-examples=examples`, but only on proc-macro crates as it seems. This issue first appeared after https://github.com/rust-lang/rust/issues/94926 was fixed.

The panic occurs here: https://github.com/rust-lang/cargo/blob/master/src/cargo/ops/cargo_compile.rs#L1600

### Steps

1. Clone [Relm4](https://github.com/Relm4/relm4)
2. Create docs for `relm4-macros` using `cargo +nightly doc -Z rustdoc-scrape-examples=examples -p relm4-macros`

This will likely work with any proc-macro crate though, as it does not even start to compile anything before cargo panics.

### Possible Solution(s)

_No response_

## Backtrace

```
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', src/tools/cargo/src/cargo/ops/cargo_compile.rs:1600:36
stack backtrace:
   0: rust_begin_unwind
             at /rustc/e7575f9670f3c837def3d186ae09366c75c7632e/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/e7575f9670f3c837def3d186ae09366c75c7632e/library/core/src/panicking.rs:142:14
   2: core::panicking::panic
             at /rustc/e7575f9670f3c837def3d186ae09366c75c7632e/library/core/src/panicking.rs:48:5
   3: <core::iter::adapters::map::Map<core::slice::iter::Iter<cargo::core::compiler::unit::Unit>, cargo::ops::cargo_compile::rebuild_unit_graph_shared::{closure#1}> as core::iter::traits::iterator::Iterator>::fold::<(), core::iter::traits::iterator::Iterator::for_each::call<cargo::core::compiler::unit::Unit, <alloc::vec::Vec<cargo::core::compiler::unit::Unit> as alloc::vec::spec_extend::SpecExtend<cargo::core::compiler::unit::Unit, core::iter::adapters::map::Map<core::slice::iter::Iter<cargo::core::compiler::unit::Unit>, cargo::ops::cargo_compile::rebuild_unit_graph_shared::{closure#1}>>>::spec_extend::{closure#0}>::{closure#0}>
   4: cargo::ops::cargo_compile::rebuild_unit_graph_shared
   5: cargo::ops::cargo_compile::create_bcx
   6: cargo::ops::cargo_compile::compile_ws
   7: cargo::ops::cargo_compile::compile
   8: cargo::ops::cargo_doc::doc
   9: cargo::commands::doc::exec
  10: cargo::cli::main
  11: cargo::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

### Version

```text
cargo 1.62.0-nightly (dba5baf 2022-04-13)
release: 1.62.0-nightly
commit-hash: dba5baf4345858c591517b24801902a062c399f8
commit-date: 2022-04-13
host: x86_64-unknown-linux-gnu
libgit2: 1.4.2 (sys:0.14.2 vendored)
libcurl: 7.80.0-DEV (sys:0.4.51+curl-7.80.0 vendored ssl:OpenSSL/1.1.1m)
os: OracleLinux 35.0.0 [64-bit]
```

### PR DESCRIPTION
**Pull request by willcrichton**
### What does this PR try to resolve?

After #10343, we have identified a clear set of conditions for whether a particular target should be considered by `-Zrustdoc-scrape-examples`. These conditions are described more clearly in #11425.

However, after some testing with complex Cargo workspaces (e.g. [wasmtime](https://github.com/bytecodealliance/wasmtime/)), I realized that the current approach of modifying the `CompileFilter` did not correctly implement this new specification. For example, a target with `doc = false` should not be scraped by default since it is not a documented unit, but the current approach would potentially include such a target for scraping.

This PR provides a new approach which I believe correctly implements the specification: 
1. `generate_targets` is called with the same parameters except the `mode` which becomes `CompileMode::Docscrape` instead of `CompileMode::Doc`. `filter_default_targets` generates the same targets for `Docscrape` as for `Doc`.
2. Inside `generate_targets`, an initial set of `Proposal`s are created. This set of proposals is extended with further proposals based on targets identified as `doc-scrape-examples = true`, or Example targets where possible.

This PR subsumes #11423, and also fixes #10571.

### How should we test and review this PR?

I have added another test `docscrape::only_scrape_documented_targets` to verify that only documented or explicitly-enabled targets are included for scraping.

r? @weihanglo

### CODE REVIEW

File: src/cargo/ops/cargo_compile/mod.rs [hunk 1]
```diff
@@ -597,6 +591,7 @@ fn generate_targets(
     package_set: &PackageSet<'_>,
     profiles: &Profiles,
     interner: &UnitInterner,
+    has_dev_units: HasDevUnits,
```
Comment (weihanglo) [inline src/cargo/ops/cargo_compile/mod.rs:594]: On no. The thirteenth parameter 😂

File: src/cargo/ops/cargo_compile/mod.rs [hunk 2]
```diff
@@ -828,6 +823,52 @@ fn generate_targets(
         }
     }
 
+    if mode.is_doc_scrape() {
```
Comment (weihanglo) [inline src/cargo/ops/cargo_compile/mod.rs:826]: Not a blocker. Do you feel it better if we extract this piece of code to a separate function doing post-process after calling `generate_targets`? Doing so avoids adding more parameters on `generate_targets`. 

Don't blindly agree with me, though 😆.

**Review by weihanglo — COMMENTED** [review]
Make sense. Thank you!

**Comment by willcrichton** [comment]
@weihanglo I share your concern that `generate_targets` is getting really complicated. I actually started to refactor it, but decided to wait.

If it's ok, I would like to merge this PR as-is to get the functionality into Cargo. After this is merged, I will file a second PR that refactors `generate_targets` into a new file `cargo_compile/target_generator.rs` that has a struct `TargetGenerator` with the mega-method factored into different documentable pieces.

File: src/cargo/ops/cargo_compile/mod.rs (hunk 2 above)
Comment (willcrichton) [reply src/cargo/ops/cargo_compile/mod.rs:826]: The issue is that this code needs to be in the middle of `generate_targets` --- it post-processes the *proposals* but not the *units*. 

In my proposed refactor, I'm going to separate out `generate_targets` into `TargetGenerator::make_proposals` and `TargetGenerator::proposals_to_units`. Then I can cleanly insert the doc-scrape-specific code after `make_proposals` as post-processing.

**Review by weihanglo — APPROVED** [review]
Thank you for taking care of these edge cases I was not aware of!

It's fine to refactor later. Really looking forward to seeing a well-organized unit generator!

**Pull request merged**